tests/integration/testvalidator.nim

```diff
@@ -0,0 +1,215 @@
+    # Uncomment to start Hardhat automatically, typically so logs can be inspected locally
+    hardhat:
+      HardhatConfig.none,
+
+    clients:
+      CodexConfigs.init(nodes=1)
+        # .debug() # uncomment to enable console log output
+        # .withLogFile() # uncomment to output log file to tests/integration/logs/<start_datetime> <suite_name>/<test_name>/<node_role>_<node_idx>.log
+        # .withLogTopics("node", "marketplace", "clock")
+        .some,
+
+    providers:
+      CodexConfigs.init(nodes=1)
+        .withSimulateProofFailures(idx=0, failEveryNProofs=1)
+        # .debug() # uncomment to enable console log output
+        # .withLogFile() # uncomment to output log file to tests/integration/logs/<start_datetime> <suite_name>/<test_name>/<node_role>_<node_idx>.log
+        # .withLogTopics("marketplace", "sales", "reservations", "node", "clock", "slotsbuilder")
+        .some
+  ):
+    let client0 = clients()[0].client
+    let expiry = 5.periods
+    let duration = expiry + 10.periods
+
+    let data = await RandomChunker.example(blocks=8)
+
+    # TODO: better value for data.len below. This TODO is also present in
+    # testproofs.nim - we may want to address it or remove the comment.
+    createAvailabilities(data.len * 2, duration)
+
+    let slotFilledSubscription = await trackSlotsFilled(marketplace)
+
+    let cid = client0.upload(data).get
+
+    let purchaseId = await client0.requestStorage(
+      cid,
+      expiry=expiry,
+      duration=duration,
+      nodes=nodes,
+      tolerance=tolerance,
+      proofProbability=1
+    )
+    let requestId = client0.requestId(purchaseId).get
+
+    check eventually(client0.purchaseStateIs(purchaseId, "started"), 
+      timeout = expiry.int * 1000)
+    
+    # just to make sure we have a mined block that separates us
+    # from the block containing the last SlotFilled event
+    discard await ethProvider.send("evm_mine")
+
+    var validators = CodexConfigs.init(nodes=2)
+      .withValidationGroups(groups = 2)
+      .withValidationGroupIndex(idx = 0, groupIndex = 0)
+      .withValidationGroupIndex(idx = 1, groupIndex = 1)
+      # .debug() # uncomment to enable console log output
+      # .withLogFile() # uncomment to output log file to:
+      # tests/integration/logs/<start_datetime> <suite_name>/<test_name>/<node_role>_<node_idx>.log
+      # .withLogTopics("validator") # each topic as a separate string argument
+    
+    failAndTeardownOnError "failed to start validator nodes":
+      for config in validators.configs.mitems:
+        let node = await startValidatorNode(config)
+        running.add RunningNode(
+          role: Role.Validator,
+          node: node
+        )
+    
+    let slotFreedSubscription = 
+      await trackSlotsFreed(requestId, marketplace)
+
+    let expectedSlotsFreed = nodes - tolerance
+    check eventually((slotsFreed.len == expectedSlotsFreed),
+      timeout=(duration - expiry).int * 1000)
```

Comment:
@markspanbroek suggest that based on the observation that `expectedSlotsFreed` should be equal to `tolerance + 1`, it might be sufficient to simply wait for the storage request to fail instead of tracking slots freed. Make sense to me, I will check it out and push changes.

Reply:
ok, so I took a mental step back. The purpose of this test is to check that validators are doing their work. We can detect that by looking at failing request, but this is actually not the most direct way of sensing that. Tracking events is. We can sense that slots are being freed and that request fails and this would not happen without properly working validator. Sensing a failing request by polling the state of the purchase should bring us to the same results, yet, the sensing path is longer and more pone to an accidental errors which does not currently seem to be completely avoidable. Errors like:

```
Exception in blockexc read loop   topics="codex blockexcnetworkpeer" tid=13304 msg="Stream Closed!"
```

or:

```
Unhandled exception in async proc, aborting topics="codex" tid=13640 msg="C:\\msys64\\home\\code\\...\\Nim\\lib\\pure\\json.nim(824, 10) `node.kind == JArray` : items() can not iterate a JsonNode of kind JNull"
```

Especially that those problems mostly happen on slow and less reliable Windows runners.

Thus, while looking for failing purchase is elegant and strips some of the code from the test, it is not the shortest and the most reliable way of testing what is supposed to be tested here.

In this spirit, and to keep the integration tests as short as possible, I decided to track the marketplace events to confirm that validator correctly detected the missing proofs.

Reply:
The first exception shouldn't really be affecting validation as it has to do with the block exchange. This could affect datasets from being downloaded, but that's doesn't seem to be an issue.

The second exception was related to when polled (http) subscriptions returned `null` instead of `[]` (possibly due to an expired filter on the client). This was fixed in https://github.com/codex-storage/nim-ethers/pull/94.

"Sensing a failing request" can be done by:
1. Checking the request state: `await marketplace.getRequest(requestId.get)` (although we should probably expose this on a REST API endpoint instead of querying the contract directly, for another PR)
2. Checking the `slotState` for all slots in the request (longer, more verbose)

If the contract is showing that a request is failed, we can trust that it is failed for certain.

Reply:
> The second exception was related to when polled (http) subscriptions returned null instead of [] (possibly due to an expired filter on the client). This was fixed in https://github.com/codex-storage/nim-ethers/pull/94.

I know it was fixed in https://github.com/codex-storage/nim-ethers/pull/94 - I spent lots of time to track that error down ;).

> If the contract is showing that a request is failed, we can trust that it is failed for certain.

It is not a matter of trust for me here, but about shortening the distance between the thing under test and where we are sensing it. Saying that, I did not consider using `marketplace.getRequest(requestId.get)`, which indeed could be a good middle ground. I think it is better to query the contract in this case, because this is the closest observable thing. Using REST API (or any logic in the client for that matter) feels to me like introducing an extra distance. If something in the client goes wrong (when for instance we get a bug there), you will be having a failing validator test even though the validator and the contract are working correctly. I like when a failing test helps me locate the problem quickly...

Reply:
> The second exception was related to when polled (http) subscriptions returned null instead of [] (possibly due to an expired filter on the client). This was fixed in https://github.com/codex-storage/nim-ethers/pull/94.

I know it was fixed in https://github.com/codex-storage/nim-ethers/pull/94 - I spent lots of time to track that error down ;).

> If the contract is showing that a request is failed, we can trust that it is failed for certain.

It is not a matter of trust for me here, but about shortening the distance between the thing under test and where we are sensing it. Saying that, I did not consider using `marketplace.getRequest(requestId.get)`, which indeed could be a good middle ground. I think it is better to query the contract in this case, because this is the closest observable thing. Using REST API (or any logic in the client for that matter) feels to me like introducing an extra distance. If something in the client goes wrong (when for instance we get a bug there), you will be having a failing validator test even though the validator and the contract are working correctly. I like when a failing test helps me locate the problem quickly...

Reply:
I can also use `await marketplace.requestState(requestId)`. This seems to be working good and this way I do not need to track events so all event tracking logic is removed.

Reply:
Done. Event tracking removed.